crates/ra_lsp_server/src/main_loop/handlers.rs

```diff
@@ -1060,3 +1061,72 @@ pub fn handle_call_hierarchy_outgoing(
 
     Ok(Some(res))
 }
+
+pub fn handle_semantic_tokens(
+    world: WorldSnapshot,
+    params: SemanticTokensParams,
+) -> Result<Option<SemanticTokensResult>> {
+    let _p = profile("handle_semantic_tokens");
+
+    let file_id = params.text_document.try_conv_with(&world)?;
+    let line_index = world.analysis().file_line_index(file_id)?;
+
+    let mut builder = SemanticTokensBuilder::default();
+
+    for h in world.analysis().highlight(file_id)?.into_iter() {
+        let type_and_modifiers: (SemanticTokenType, Vec<SemanticTokenModifier>) =
+            h.tag.conv_with(&world);
+        let (token_type, token_modifiers) = type_and_modifiers.conv_with(&world);
+        builder.push(h.range.conv_with(&line_index), token_type, token_modifiers);
+    }
+
+    let tokens = SemanticTokens { data: builder.build(), ..Default::default() };
+
+    Ok(Some(tokens.into()))
+}
+
+#[derive(Default)]
+struct SemanticTokensBuilder {
+    prev_line: u32,
+    prev_char: u32,
+    data: Vec<SemanticToken>,
+}
+
+impl SemanticTokensBuilder {
```

Comment:
A straight forward port of the typescript version but without the support for edits.